File: Difficulty: Easy/Find length of Loop/find-length-of-loop.cpp
/*
class Node {
 public:
    int data;
    Node *next;

    Node(int x) {
        data = x;
        next = NULL;
    }
};

*/
class Solution {
public:
    int lengthOfLoop(Node *head) {
        Node* slow = head;
        Node* fast = head;
        
        while (fast != NULL && fast->next != NULL) {
            slow = slow->next;
            fast = fast->next->next;
            
            if (fast == slow) {
                // Loop detected, now count the length
                int len = 1;
                Node* temp = slow->next;
                while (temp != slow) {
                    len++;
                    temp = temp->next;
                }
                return len;  // Return as soon as the loop is counted
            }
        }
        // No loop found
        return 0;
    }
};

File: Difficulty: Easy/Find length of Loop/find-length-of-loop.py
'''
class Node:
    def __init__(self, data): 
        self.data = data
        self.next = None
'''
class Solution:
    def lengthOfLoop(self, head):
        #code here
        slow = head
        fast = head
    
        # Step 1: Traverse the list to detect a loop
        while fast is not None and fast.next is not None:
            # Move slow one step
            slow = slow.next     
            # Move fast two steps
            fast = fast.next.next  
    
            # Step 2: If the slow and fast
            # pointers meet, there is a loop
            if slow == fast:
                # Initialize the loop length
                length = 1  
                 # Move fast one step
                fast = fast.next 
    
                # Step 4: Initialize a counter
                # and traverse using the fast pointer
                while slow != fast:
                    # Move fast one step
                    fast = fast.next  
                    # Increment the counter
                    length += 1  
    
                # Step 6: Return the 
                # length of the loop
                return length
    
        # Step 3: If the fast pointer
        # reaches the end, there is no loop
        return 0  